stm32: Add support for STM32N6xx MCUs and two N6 boards

dpgeorge opened the pull request:
### Summary

This PR adds preliminary support for ST's new STM32N6xx MCUs.

Supported features of this MCU so far are:
- basic clock tree initialisation, running at 800MHz
- fully working USB
- mboot support (required, because there's no internal flash)
- XSPI in memory-mapped mode
- `machine.Pin`
- `machine.UART`
- SD card
- filesystem
- ROMFS
- WiFi and BLE via cyw43-driver (SDIO and UART backend)

Supported boards:
- NUCLEO_N657X0
- ~~STM32N6570_DK~~ Edit: this board is no longer supported, can be added later if needed
- OPENMV_N6

Note that the N6 does not have internal flash, and has some tricky boot sequence, so using a custom bootloader (mboot) is almost a necessity.

~~The ST CMSIS and HAL files are added verbatim here, but will eventually be moved into `stm32lib`.~~  Edit: N6 CMSIS and HAL files are now in `stm32lib`.

OpenMV have generously sponsored the development of this port.

### Testing

This PR has been tested on the two N6 boards that are added here.

dpgeorge committed:
stm32: Add support for STM32N6xx MCUs.

This commit adds preliminary support for ST's new STM32N6xx MCUs.

Supported features of this MCU so far are:
- basic clock tree initialisation, running at 800MHz
- fully working USB
- XSPI in memory-mapped mode
- machine.Pin
- machine.UART
- RTC and deepsleep support
- SD card
- filesystem
- ROMFS
- WiFi and BLE via cyw43-driver (SDIO backend)

Note that the N6 does not have internal flash, and has some tricky boot
sequence, so using a custom bootloader (mboot) is almost a necessity.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/boards: Add board support files for N6.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/boards/NUCLEO_N657X0: Add new board definition files.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/boards/make-pins.py: Support up to GPIO-O.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/boards/OPENMV_N6: Add new board definition files.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/mboot: Add support for STM32N6xx MCUs.

Works in the usual USB DFU mode, and can program external SPI flash.  It
will enable XSPI memory-mapped mode before jumping to the application
firmware in the external SPI flash.

Signed-off-by: Damien George <damien@micropython.org>

iabdalkader committed:
stm32/main: Disable D-cache when debugging N6.

See ST Errata ES0620 - Rev 0.2 section 2.1.2.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented on ports/stm32/boards/OPENMV_N6/board.c:
```diff
@@ -0,0 +1,261 @@
+    // TODO: if (HAL_PWREx_ConfigSupply(PWR_EXTERNAL_SOURCE_SUPPLY ) != HAL_OK)
+    //xspi_flash_init();
+}
+
```
```suggestion

#define OMV_BOOT_MAGIC_ADDR   (0x3401FFFCU)
#define OMV_BOOT_MAGIC_VALUE  (0xB00710ADU)

void board_enter_bootloader(void) {
    *((uint32_t *) OMV_BOOT_MAGIC_ADDR) = OMV_BOOT_MAGIC_VALUE;
    SCB_CleanDCache();
    NVIC_SystemReset();
}
```
To enter our bootloader.

dpgeorge replied:
Do you want to have the OPENMV_N6 board definition included in MicroPython's stm32 port (like with the OPENMV_AE3)?  If so, it should work with mboot.  But probably also a good idea to work with the OpenMV bootloader, which means adding this code.

Related: how did you choose address 0x3401FFFC?  That's part way through SRAM1, in the FLEXRAM area.  You'd need to make sure that isn't cleared on reset like the rest of SRAM1/2, and also make sure it's not overwritten by the bootloader.

iabdalkader replied:
> Do you want to have the OPENMV_N6 board definition included in MicroPython's stm32 port (like with the OPENMV_AE3)? If so, it should work with mboot. But probably also a good idea to work with the OpenMV bootloader, which means adding this code.

Either way is fine with me. If mboot support is kept, I can always use `MP_CONFIGFILE` to override and not build mboot when building our firmware.

> Related: how did you choose address 0x3401FFFC? That's part way through SRAM1, in the FLEXRAM area. You'd need to make sure that isn't cleared on reset like the rest of SRAM1/2, and also make sure it's not overwritten by the bootloader.

Good point. The bootloader uses the 128K (or 64K) DTCM for its memory (heap, stack etc..), so this address is typically `0x2001FFFCU` (last word of bootloader's memory) and it's the same for almost all boards. However, for some reason, this doesn't work on the N6, so I just used SRAM1. Note that SRAM1 and SRAM2 don't seem to get erased on reset, otherwise this wouldn't work, but it would still be better to use DTCM for this, but it's not working.

iabdalkader replied:
Oh I think I just needed a DSB, seems the write might be buffered. The following works too:

```C
#define OMV_BOOT_MAGIC_ADDR   (0x3001FFFCU)
#define OMV_BOOT_MAGIC_VALUE  (0xB00710ADU)

void board_enter_bootloader(void) {
    *((uint32_t *) OMV_BOOT_MAGIC_ADDR) = OMV_BOOT_MAGIC_VALUE;
    __DSB();
    NVIC_SystemReset();
}
```

iabdalkader replied:
@dpgeorge There seems to be some issue with DTCM, I can't access any address above ~0x1000. This causes a fault on boot when it tries accesses the magic number address (oddly enough, it doesn't always happen). Perhaps it has something to do with security config, flexram, clocks or something else that gets enabled by the main firmware, but even from the main firmware I still can't access this memory from gdb.  Anyway let's please keep the original boot address (with cache invalidate).

dpgeorge committed:
drivers: Support special QSPI direct-read protocol.

This is useful for interfaces that stay in memory-mapped mode by default.
They can implement this method with a simple `memcpy()`.

Signed-off-by: Damien George <damien@micropython.org>

iabdalkader commented on ports/stm32/boards/OPENMV_N6/mpconfigboard.mk:
```diff
@@ -12,15 +12,26 @@ STM32_N6_HEADER_VERSION = 2.3
+MICROPY_PY_NETWORK_CYW43 = 1
+MICROPY_PY_SSL = 1
+MICROPY_SSL_MBEDTLS = 1
+MICROPY_VFS_LFS2 = 1
```
I need to be able to disable mboot `USE_MBOOT ?= 1` and also we never use LFS2, just fat.

iabdalkader replied:
But what if you don't want to build mboot, yet still want a flash-based image? Can `USE_MBOOT=1` define something like `MICROPY_FLASH_BASED=1`? For example:
```
USE_MBOOT ?= 1
MICROPY_FLASH_BASED ?= $(USE_MBOOT)
```

 This way I can define `USE_MBOOT=0 MICROPY_FLASH_BASED=1` to get a flash-based image without mboot.

iabdalkader commented on ports/stm32/usbd_conf.c:
```diff
@@ -220,6 +229,23 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd) {
+        LL_AHB5_GRP1_EnableClockLowPower(LL_AHB5_GRP1_PERIPH_OTGPHY1);
+
+        // Select 24MHz clock.
+        MODIFY_REG(USB1_HS_PHYC->USBPHYC_CR, USB_USBPHYC_CR_FSEL, 2 << USB_USBPHYC_CR_FSEL_Pos);
```
There's a proper init sequence for this in the examples in the HAL, if you want to take a look. I think it's more or less the same, but there were some delays, more force_reset/release etc...

Also, could you please add this? I use it in HS mode.

```C
diff --git a/ports/stm32/usbdev/class/inc/usbd_cdc_msc_hid.h b/ports/stm32/usbdev/class/inc/usbd_cdc_msc_hid.h
index 3a87896b4..5906f95e0 100644
--- a/ports/stm32/usbdev/class/inc/usbd_cdc_msc_hid.h
+++ b/ports/stm32/usbdev/class/inc/usbd_cdc_msc_hid.h
@@ -11,7 +11,7 @@
 
 // Work out if we should support USB high-speed device mode
 #if MICROPY_HW_USB_HS \
-    && (!MICROPY_HW_USB_HS_IN_FS || defined(STM32F723xx) || defined(STM32F733xx))
+    && (!MICROPY_HW_USB_HS_IN_FS || defined(STM32F723xx) || defined(STM32F733xx) || defined(STM32N6))
 #define USBD_SUPPORT_HS_MODE (1)
 #else
 #define USBD_SUPPORT_HS_MODE (0)
```

dpgeorge replied:
I've added the `USBD_SUPPORT_HS_MODE` option as above.

But testing it with #15909 shows that there is data corruption when HS is enabled.  That needs investigation.

dpgeorge committed:
lib/stm32lib: Update library for N6 v1.1.0.

Changes in this new library version are:
- Add N6 HAL at v1.1.0.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
lib/libm_dbl: Support FLT_EVAL_METHOD == 16.

That's almost the same as FLT_EVAL_METHOD == 0, but indicates the
presence of _Float16_t support.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I finally got the filesystem working.  Needed to make sure all relevant code and data structures to erase/write SPI flash is in RAM, and that interrupts are fully disabled during erase/write (obvious but tricky to do).

iabdalkader commented:
I pulled this PR and tested it again. Things still seem to be working, and I noticed some improvements (like no longer calling `xspi_init()` from ioctl). However, I still couldn’t get the storage working. I get a hardfault when trying to initialize the XSPI PSRAM after `storage_init()` has been called. I did move the drivers/IRQs to ITCM, following what's done here, but I think it's a different issue, maybe because because it remaps XSPI back in 111 mode, when the bus is released, which is not what we use initially.

As for CYW43, I just skipped testing it altogether, as it's still using SPI. Other than that, I don’t think this is ready to coexist with older MCUs yet, for example, it comments out a large chunk of powerctrl.c, so we should probably wait for now.

iabdalkader commented on ports/stm32/boards/common_n6_flash.ld:
```diff
@@ -0,0 +1,83 @@
+
+        KEEP(*(.isr_vector)) /* Startup code */
+
+        /* These functions need to run from ram to enable uart
```
I assume this is because IRQs are being disabled for too long? If that’s the case, would it help to re-enable IRQs after each block erase/write operation, rather than keeping them disabled across all blocks?  If I’m not wrong, this is how it's done in the mimxrt and rp2 ports, and it might help avoid the need to move all those functions and their data to RAM.

dpgeorge replied:
Actually, the UART functions don't need to go in RAM anymore.  I've fully disabled IRQs during XSPI flash erase/write.  It didn't seem to work if I just raised the IRQ priority level.

And the code in `spibdev.c` does enable IRQs after each block erase.

dpgeorge commented:
I nearly have 8-8-8 octal DTR mode working with the filesystem.  Just a few things to tidy up...

dpgeorge commented:
> I have it working here, but I use HAL drivers

Yes, thanks, I was studying that code (you use `HAL_XSPI_INSTRUCTION_8_BITS` when I think it should be `HAL_XSPI_INSTRUCTION_16_BITS`?).

iabdalkader commented:
Fantastic, thanks! I'll give it another try as soon as I can and let you know how it goes.

> So if `memcpy` lives in XSPI then, although it will work, it will be slower due to fetching instructions from the `memcpy` code in XSPI and also the source data from a different location in XSPI. Thus it's best to put `memcpy` in RAM.

The RW filesystem doesn't really need to be fast, because there are two other fast options: ROMFS and SDCARD. However, will it really be that slow? The N6 models are basically relocatable binaries (data + code) stored in XSPI. I don't see any difference between running it in XIP mode (code in XSPI) and running it in copy mode (code copied to RAM). I think the icache does a pretty good job of hiding this.

Anyway, I'll try it again. But if it doesn't work, I can't think of any other fix for this (short of placing the entire libc in memory, if that is indeed the issue).

iabdalkader commented on ports/stm32/boards/common_n6_flash.ld:
```diff
@@ -0,0 +1,83 @@
+/* define output sections */
+SECTIONS
+{
+    .isr_vector :
+    {
+        . = ALIGN(4);
+        __isr_vector_ram_start = .;
+
+        KEEP(*(.isr_vector)) /* Startup code */
+
+        /* These functions need to run from ram to enable uart
+           reception during flash erase/write operations.
+           Defining them here ensures they're copied from
+           flash (in main.c) along with the isr_vector above.
+        */
+        . = ALIGN(4);
+        *(.text.mp_sched_keyboard_interrupt)
+        *(.text.pendsv_schedule_dispatch)
+        *(.text.storage_systick_callback)
+        *(.text.SysTick_Handler)
+        *(.text.uart_irq_handler)
+        *(.text.UART*_IRQHandler)
+        *(.text.USART*_IRQHandler)
+        *(.text.HAL_GetTick)
+        *drivers/memory/spiflash.o(.text* .rodata*)
+        *boards*(.rodata.spiflash_config*)
+
+        . = ALIGN(4);
```
> even some HAL functions (which are inline... but the compiler still makes them non-inline).

Could it be the `-Os` ?

> There's already asm code in resethandler_m3_iram.s that does the copy. The code in main.c that copies to RAM was already there before N6 support.

So code in main is not really needed? Just for my own understanding, why do we need so much code (including HAL GPIO functions) in RAM? Is it needed when waking up from STOP mode? I assume standby doesn't really need that because it resets on wakeup.

dpgeorge replied:
I've now cleaned this IRAM stuff up, and there's now only a single RAM section that needs to be copied.  That copy is done in asm in `resethandler_m3_iram.s`.




> Just for my own understanding, why do we need so much code (including HAL GPIO functions) in RAM? Is it needed when waking up from STOP mode? I assume standby doesn't really need that because it resets on wakeup.

Code in RAM is needed for two things:
- Writing to flash, which needs all the SPI flash logic from drivers/memory/spiflash.o and XSPI code
- Waking from **standby**!  The N6 has SRAM1 retention during standby and a special SYSCFG register that you can set to tell the MCU where to resume when waking from standby.  This allows it to wake up much faster from standby by executing code straightaway from SRAM1, instead of going through the bootloaders.  So there needs to be enough code in RAM to be able to act like a mini bootloader that can do just enough to enable XSPI memory mapped mode.

dpgeorge replied:
This is for deepsleep!  It doesn't do a reboot after waking but rather executes from retention SRAM1.  That's why it's so complicated to make it work.

dpgeorge replied:
At the very least, the SPI flash needs to be reset to take it out of 888 DTR mode, if we want to run the ST ROM bootloader.

kwagyeman replied:
Ah... do you have the reset fix on the board?

Removing these two parts fixes an issue with reset not resetting the flash:

![image](https://github.com/user-attachments/assets/6b1d87a5-b138-4696-9d00-0c0c9093266d)

After they are removed the flash should reset correctly.

kwagyeman replied:
Oh, okay, then I already updated your board to 10nF. So, it's not an issue with the reset line.

dpgeorge replied:
> Is `spiflah.o` really needed in RAM or is it a speed optimization? Could we get away with just the XSPI code as it does the actual erase/write?

Yes it's needed.  That code has all the logic to set WREN and wait for the erase to complete, for example.

Also, as mentioned, if the SPI flash is in DTR mode then it won't work with the ST ROM bootloader if the SPI flash is not fully reset upon waking from standby/deepsleep.  And I don't think it is fully reset when waking from standby.  If so, that would need code in SRAM1 in retention mode to reset it, before doing an NVIC reset (if that even works, I couldn't get it to bounce into the ST ROM bootloader).

iabdalkader replied:
> if the SPI flash is in DTR mode then it won't work with the ST ROM bootloader if the SPI flash is not fully reset upon waking from standby/deepsleep

The boards are designed such that they reset the flash on NVIC reset (same should apply for devkits, we're following the reference design here), so it should fully reset the flash if an NVIC reset occurs. This feature is crucial for them to function, otherwise `machine.reset()` wouldn't work. So normally you just need to ensure an NVIC reset occurs somehow. How about using the address of something like this instead? Would this simplify things a bit?

```C
__attribute__((noreturn, section(".ram_function"))) void board_reset(void) {
    // NVIC_SystemReset doesn't get inlined here.
    SCB->AIRCR  = (uint32_t)((0x5FAUL << SCB_AIRCR_VECTKEY_Pos) |
                             (SCB->AIRCR & SCB_AIRCR_PRIGROUP_Msk) |
                             SCB_AIRCR_SYSRESETREQ_Msk);
    __DSB();
    for (;;) {
        __NOP();
    }
}
```

>  Is that really such an issue? We could try to optimise that down a bit. Eg maybe it doesn't need the app ISR moved there (although I'd suggest doing that regardless, for IRQ performance reasons.)

Probably not, it's not the memory size I'm concerned about, in fact I put all of that code in ITCM which is not used at all, it's just that I use a shared linker script so I'm trying to figure out the bare minimum customization needed for the N6.

kwagyeman commented:
Damien, can you enable UART3/4 on the N6? PA11/PA12 and PB10/PB11?

kwagyeman commented:
About the SD card interface. If SD_RESET is pulled high, by default, UHS-1 is available. If it's low by default, then it's not.

iabdalkader commented on ports/stm32/boards/OPENMV_N6/mpconfigboard.h:
```diff
@@ -0,0 +1,168 @@
+#define MICROPY_HW_ENABLE_DAC       (0)
+#define MICROPY_HW_ENABLE_USB       (1)
+#define MICROPY_HW_ENABLE_SDCARD    (0) // TODO enable
+#define MICROPY_PY_PYB_LEGACY       (0)
```
```c
#define MICROPY_FATFS_EXFAT         (1)
```
We use exfat for all boards.

dpgeorge replied:
Now enabled on OPENMV_N6.

iabdalkader commented:
Re waking up from deepsleep, I got the following working (reusing the same trick I used for NV, by rigging a memory ISR):

```C
static char _boot_mem[1024] __attribute__((aligned(1024)));

__attribute__((naked, noreturn, section(".ram_function"))) void ram_reset(void) {
    // NVIC_SystemReset doesn't get inlined here.
    SCB->AIRCR  = (uint32_t)((0x5FAUL << SCB_AIRCR_VECTKEY_Pos) |
                             (SCB->AIRCR & SCB_AIRCR_PRIGROUP_Msk) |
                             SCB_AIRCR_SYSRESETREQ_Msk);
    __DSB();
    for (;;) {
        __NOP();
    }
}

void board_early_init(void) {
    HAL_PWREx_EnableTCMRetention();
    HAL_PWREx_EnableTCMFLXRetention();

    uint32_t *boot_mem = (uint32_t *)_boot_mem;
    boot_mem[0] = (uint32_t)(_boot_mem + 1024);
    boot_mem[1] = ((uint32_t)&ram_reset) | 1;
    SCB_CleanDCache_by_Addr((uint32_t *)_boot_mem, 32);

    SYSCFG->INITSVTORCR = (uint32_t) boot_mem;
    (void) SYSCFG->INITSVTORCR;
}
```

It causes an NVIC reset on wakeup, and the bootloader runs again, with flash out of 888. Though, I'm not 100% happy with it, there should be an easier way to get the old STM behavior (NVIC reset on wakeup). I'll ask around, but if this is the only way, it's still considerably less complex than the asm code, `iram_bootloader_reset`, etc... 

Note that this doesn't have to be used in MicroPython, it's self contained in board.c, just need an option to not change `INITSVTORCR`.

dpgeorge commented on ports/stm32/sdcard.c:
```diff
@@ -40,7 +40,7 @@
 #if MICROPY_HW_ENABLE_SDCARD || MICROPY_HW_ENABLE_MMCARD
 
-#if defined(STM32F7) || defined(STM32H5) || defined(STM32H7) || defined(STM32L4)
+#if defined(STM32F7) || defined(STM32H5) || defined(STM32H7) || defined(STM32L4) || defined(STM32N6)
```
Yes, I just enabled low power clock for everything, because I was having so many issues with WFI!

Do you think it's better to enable them individually when needed?

Well, at least for things like RAM they need to be enabled in main.

iabdalkader replied:
> Do you think it's better to enable them individually when needed?

I'm not sure what the best approach is, but that's what I do in our firmware (enable them on demand). I've added the corresponding `CLK_SLEEP_ENABLE` to every `CLK_ENABLE`.

Re RAMs, I agree they need to be enabled early on. I enable some RAMs in the reset handler (I think those are used by the bootloader), and others in `SystemInit`.

dpgeorge replied:
I've enabled some sleep clocks unconditionally in main (SPI, UART, I2C, memory), but left others to enable them explicitly when needed.

I think if you enable a peripheral, you probably want it to stay awake during WFI/WFE.

iabdalkader commented:
Just noticing a weird issue with WiFi, it seems to work fine with debug builds, but with non-debug builds only scan works.. Does this sound familiar? Perhaps I'm missing something or the memory layout changes and breaks something.

dpgeorge committed:
stm32/lwip_inc: Increase lwIP memory on N6.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented on ports/stm32/boards/OPENMV_N6/mpconfigboard.h:
```diff
@@ -0,0 +1,176 @@
+#define MICROPY_HW_USB_HS           (1)
+#define MICROPY_HW_USB_HS_IN_FS     (1)
+#define MICROPY_HW_USB_MAIN_DEV     (USB_PHY_HS_ID)
+#define MICROPY_HW_USB_MSC          (0)
```
MSC now enabled on OPENMV_N6, and I fixed it so it works with internal flash (needed to increase MSC buffer size from 2k to 4k because that's the flash erase size).

kwagyeman commented:
It needs 32 on the N6. Ibrahim got it working with that.

dpgeorge commented:
> I didn't test ping, just a simple scan + connect to AP.

For me, scan always works (regardless of mem align).  Connect to AP only works with mem align 8 or more.  Ping never works.

We probably need to understand why mem alignment matters, because mem align of 4 should be enough.  (Actually I'd be surprised if it even needs to be 4, it should work at 1 because ARM allows unaligned accesses.)

iabdalkader commented:
> For me, scan always works (regardless of mem align). Connect to AP only works with mem align 8 or more. Ping never works.

Yes, the exact same thing here, except I didn't test ping.

> We probably need to understand why mem alignment matters, because mem align of 4 should be enough

I'm wondering if alignment always mattered and we were just lucky with other boards, or if it's just the N6. It could have something to do with cache, might be worth testing again with 4 + dcache disabled.

iabdalkader commented:
@dpgeorge Forgot to mention that compiling lwip objects with `-O2` (custom rule in `extmod.mk`) also makes AP connect work for me. This may indicate that it has something to do with memory layout.

iabdalkader commented:
Any updates on the LWIP issue? I think this is pretty much ready to merge (at least in our fork), but we probably should wait for the LWIP fix.

dpgeorge commented:
Building without -Og, but with alignment of 8, incoming ping packets fail their checksum check.

I can't load that lwip issue you link to (it times out) but if the date of that is really 2011 then that's a long time ago and it should be fixed by now?  Also strange that the issue is not seen on other boards using lwIP.

iabdalkader commented:
> I can't load that lwip issue you link to (it times out) but if the date of that is really 2011 then that's a long time ago and it should be fixed by now? Also strange that the issue is not seen on other boards using lwIP.

That's weird, it was working just yesterday. Anyway, building the checksum function with -O0 or -Og just introduced an extra delay which masked the issue. The OP eventually found out that it was his DMA code corrupting frames. We're likely looking at the same issue, and if we assume it doesn't exist on other ports or with other MCUs then it has to be the new SDIO code.

iabdalkader commented:
> I was investigating the `lwip_standard_chksum()` function in lwIP because it was returning strangely incorrect values

Should be easy to test the function with known input/output. I'll give it a shot. FWIW I quickly tested `gcc version 14.2.1 20241119` and it doesn't fix the issue, nor does changing the checksum algorithm (tried 1 to 3). Not sure how to proceed, can we just add an -Og for that checksum file for N6? Everything else seems to be working perfectly fine. I'm leaning towards not gcc bug because we have a ton of optimize, image processing code that works fine. Note that we always build our firmware with `-O2`.

iabdalkader commented:
@dpgeorge This produces different results with/without -Og, you were right something is definitely broken...
```diff
diff --git a/ports/stm32/modpyb.c b/ports/stm32/modpyb.c
index 3d8696e3c..d389bb5cd 100644
--- a/ports/stm32/modpyb.c
+++ b/ports/stm32/modpyb.c
@@ -133,6 +133,50 @@ static mp_obj_t pyb_country(size_t n_args, const mp_obj_t *args) {
 }
 static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_network_country_obj, 0, 1, pyb_country);
 #endif
+#include "lwip/inet_chksum.h"
+#include "lwip/def.h"
+
+static u16_t __attribute__((optimize("Og")))  lwip_standard_chksum(const void *dataptr, int len) {
+  u32_t acc;
+  u16_t src;
+  const u8_t *octetptr;
+
+  acc = 0;
+  /* dataptr may be at odd or even addresses */
+  octetptr = (const u8_t *)dataptr;
+  while (len > 1) {
+    /* declare first octet as most significant
+       thus assume network order, ignoring host order */
+    src = (*octetptr) << 8;
+    octetptr++;
+    /* declare second octet as least significant */
+    src |= (*octetptr);
+    octetptr++;
+    acc += src;
+    len -= 2;
+  }
+  if (len > 0) {
+    /* accumulate remaining octet */
+    src = (*octetptr) << 8;
+    acc += src;
+  }
+  /* add deferred carry bits */
+  acc = (acc >> 16) + (acc & 0x0000ffffUL);
+  if ((acc & 0xffff0000UL) != 0) {
+    acc = (acc >> 16) + (acc & 0x0000ffffUL);
+  }
+  /* This maybe a little confusing: reorder sum using lwip_htons()
+     instead of lwip_ntohs() since it has a little less call overhead.
+     The caller must invert bits for Internet sum ! */
+  return lwip_htons((u16_t)acc);
+}
+
+static mp_obj_t pyb_checksum(mp_obj_t data) {
+    mp_buffer_info_t rbuf;
+    mp_get_buffer_raise(data, &rbuf, MP_BUFFER_READ);
+    return mp_obj_new_int(lwip_standard_chksum(rbuf.buf, rbuf.len));
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(pyb_checksum_obj, pyb_checksum);
 
 static const mp_rom_map_elem_t pyb_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_pyb) },
@@ -260,6 +304,7 @@ static const mp_rom_map_elem_t pyb_module_globals_table[] = {
     #if MICROPY_HW_HAS_LCD
     { MP_ROM_QSTR(MP_QSTR_LCD), MP_ROM_PTR(&pyb_lcd_type) },
     #endif
+    { MP_ROM_QSTR(MP_QSTR_checksum), MP_ROM_PTR(&pyb_checksum_obj) },
 };
 
 static MP_DEFINE_CONST_DICT(pyb_module_globals, pyb_module_globals_table);
```

dpgeorge commented on ports/stm32/boards/OPENMV_N6/mpconfigboard.h:
```diff
@@ -0,0 +1,175 @@
+
+#define MICROPY_BOARD_ENTER_BOOTLOADER board_enter_bootloader
+#define MICROPY_BOARD_EARLY_INIT    board_early_init
+#define MICROPY_BOARD_LEAVE_STANDBY board_leave_standby
```
I was trying to be consistent with `MICROPY_BOARD_STARTUP` and `MICROPY_BOARD_EARLY_INIT` etc.

The problem with having the `;` part of the macro is that the use of the macro doesn't end in a `;` and hence isn't really well formatted C code, and can give problems with the auto-formatter (uncrustify).

kwagyeman commented:
I can verify this issue is present on 13.2.rel1 and 13.3.rel1. O2 seems to fix the issue.

kwagyeman commented on ports/stm32/boards/OPENMV_N6/board.ld:
```diff
@@ -0,0 +1,12 @@
+
+MEMORY
+{
+    EXT_FLASH_ROMFS (rx) : ORIGIN = 0x70200000, LENGTH = 14M
```
24M?

dpgeorge replied:
Now 24M.

kwagyeman commented on ports/stm32/lwip_inc/lwipopts.h:
```diff
@@ -10,9 +10,24 @@
 
+// Increase memory for lwIP to get better performance.
+#if defined(STM32N6)
+#define MEM_SIZE                        (16 * 1024)
```
Make larger? We have 1GB/s Ethernet onboard.

iabdalkader replied:
I prefer if we use the config I made for mimxrt and alif, back then I checked it against the docs and every option is carefully calculated and it performs very well:
```
#define MEM_SIZE                        (16 * 1024)
#define TCP_MSS                         (1460)
#define TCP_OVERSIZE                    (TCP_MSS)
#define TCP_WND                         (8 * TCP_MSS)
#define TCP_SND_BUF                     (8 * TCP_MSS)
#define TCP_SND_QUEUELEN                (2 * (TCP_SND_BUF / TCP_MSS))
#define TCP_QUEUE_OOSEQ                 (1)
#define MEMP_NUM_TCP_SEG                (2 * TCP_SND_QUEUELEN)
```
Could increase the MEM_SIZE if you like.

dpgeorge commented:
I can confirm that using `-fdisable-rtl-loop2_doloop` fixes the issue.  So I suggest using that flag on compilers known to have the bug.

iabdalkader commented on ports/stm32/boards/OPENMV_N6/mpconfigboard.h:
```diff
@@ -0,0 +1,175 @@
+#define MICROPY_HW_BDEV_SPIFLASH                (&spi_bdev)
+#define MICROPY_HW_BDEV_SPIFLASH_EXTENDED       (&spi_bdev)
+#define MICROPY_HW_BDEV_SPIFLASH_CONFIG         (&spiflash_config)
+#define MICROPY_HW_BDEV_SPIFLASH_OFFSET_BYTES   (16 * 1024 * 1024)
+#define MICROPY_HW_BDEV_SPIFLASH_SIZE_BYTES     (16 * 1024 * 1024)
```
@dpgeorge If I understand correctly, this is the offset at which the flash filesystem (writable fs) starts and its size, correct? If so, this needs to be 4MBs and 4MBs, and the romfs offset in the linker script needs to be at 8MBs and size 24MBs. This is the current flash layout we use:
4MBs -> firmware
4MBs-> RW FS
24MBs-> ROMFS

I can change that, if there's a good reason to put the rw FS last?

kwagyeman replied:
Seconding Ibrahim here. We've shipped alpha N6 boards to a few customers. Would prefer Micropython matches OpenMV's flash layout so they don't have to update their bootloaders.

dpgeorge replied:
> I can change that, if there's a good reason to put the rw FS last?

It doesn't matter the order.  I will update this to match your settings.

dpgeorge replied:
Now updated.

iabdalkader commented:
> `-fdisable-rtl-loop2_doloop`

This option generates one warning/note per file, that seems impossible to suppress.

dpgeorge commented:
> This option generates one warning/note per file, that seems impossible to suppress.

Yes I saw that.  I don't think there's much choice, except to update to a non-broken compiler.

I will make sure `-fdisable-rtl-loop2_doloop` is only enabled for gcc versions that are known broken.

dpgeorge committed:
stm32/spi: Fail spi_init if pins can't be configured.

Follows the UART and I2C drivers.

Signed-off-by: Damien George <damien@micropython.org>

iabdalkader commented:
Note GCC 14.3.1 is out and it does seem to fix the issue. It adds a bit of text to the firmware size, not sure why, but that's okay for us at least. I suggest making Makefile fail if an older toolchain is used (For example, https://github.com/openmv/openmv/blob/master/common/check_toolchain.mk)

dpgeorge committed:
stm32/stm32.mk: Error out if compiling for cortex-m55 on old gcc.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> Note GCC 14.3.1 is out and it does seem to fix the issue.

OK, good.  I've tested it as well and it does fix the problem.



> I suggest making Makefile fail if an older toolchain is used

OK, I've made an addition for this, similar to what you suggested.

dpgeorge commented:
Merged!

kwagyeman commented:
@dpgeorge Amazing! Looking forward to Ethernet and ADC PRs now!

anchung-chen commented:
1: For the ADC to work on STM32N6, the RIF(Resource Isolation Framework) attributes of ADC  should be set in main.c
    HAL_RIF_RISC_SetSlaveSecureAttributes(RIF_RISC_PERIPH_INDEX_ADC12 , RIF_ATTRIBUTE_SEC | RIF_ATTRIBUTE_PRIV);

2: The current clock frequency of ADC is 100MHz (in powerctrlboot.c), but the sample time of ADC is only 11.5 clock (in adc.c) which is far too small.

3: Also the calling of ADC calibration function is missing for STM32N6 in adc.c

dpgeorge commented:
@anchung-chen thanks for the hints!  I've now fixed N6 ADC in PR #18009.
